aptos-move/aptos-vm/src/verifier/event_validation.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::move_vm_ext::SessionExt;
+use aptos_framework::RuntimeModuleMetadataV1;
+use move_binary_format::{
+    access::ModuleAccess,
+    errors::{Location, PartialVMError, VMError, VMResult},
+    file_format::{
+        Bytecode,
+        SignatureToken::{Struct, StructInstantiation},
+    },
+    CompiledModule,
+};
+use move_core_types::{
+    account_address::AccountAddress, language_storage::ModuleId, vm_status::StatusCode,
+};
+use std::collections::HashSet;
+
+const EVENT_MODULE_NAME: &str = "event";
+const EVENT_EMIT_FUNCTION_NAME: &str = "emit";
+
+fn metadata_validation_err(msg: &str) -> Result<(), VMError> {
+    Err(metadata_validation_error(msg))
+}
+
+fn metadata_validation_error(msg: &str) -> VMError {
+    PartialVMError::new(StatusCode::CONSTRAINT_NOT_SATISFIED)
+        .with_message(format!("metadata and code bundle mismatch: {}", msg))
+        .finish(Location::Undefined)
+}
+
+/// Validate resource group metadata on modules one by one:
+/// * Extract the event metadata
+/// * Verify all changes are compatible upgrades (existing event attributes cannot be removed)
+pub(crate) fn validate_module_events(
+    session: &mut SessionExt,
+    modules: &[CompiledModule],
+) -> VMResult<()> {
+    for module in modules {
+        let mut new_event_structs =
+            if let Some(metadata) = aptos_framework::get_metadata_from_compiled_module(module) {
+                extract_event_metadata(&metadata)?
+            } else {
+                HashSet::new()
+            };
+
+        // Check all the emit calls have the correct struct with event attribute.
+        validate_emit_calls(&new_event_structs, module)?;
+
+        let original_event_structs =
+            extract_event_metadata_from_module(session, &module.self_id())?;
+
+        for member in original_event_structs {
+            // Fail if we see a removal of an event attribute.
+            if !new_event_structs.remove(&member) {
+                metadata_validation_err("Invalid change in event attributes")?;
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Validate all the `0x1::event::emit` calls have the struct defined in the same module with event
+/// attribute.
+pub(crate) fn validate_emit_calls(
+    event_structs: &HashSet<String>,
+    module: &CompiledModule,
+) -> VMResult<()> {
+    for fun in module.function_defs() {
```

Comment:
How expensive is it if a module has a lot of emit functions? Gas is not currently charged for verifier work. This is just limited to module publishing but just in case this is expensive enough

Reply:
it should be included in the base fee